### PR TITLE
[4/N] Quantization Refactor: AWQ schemes and Kernel call and weight init split

### PR DESCRIPTION
**Pull request by Alisehen**
## Motivation

Add schemes to awq instead of storing all classes in a single file, and split kernel call and weight init. Follow up to https://github.com/sgl-project/sglang/pull/17503.
Images and motivation for this PR can be viewed in our roadmap: https://github.com/sgl-project/sglang/issues/15194.

## Modifications
 Refactored AWQ to align with the scheme-based quantization structure used by modelslim and compressed_tensors.

  Moved AWQ implementations out of the monolithic quantization/awq.py into the new package under quantization/awq/, with scheme implementations split into quantization/awq/schemes/.

  Added get_linear_scheme and get_moe_scheme to awq/awq.py so linear and MoE paths select concrete schemes explicitly.

  Unified AWQ quant methods into thin wrappers that delegate to layer.scheme, matching the compressed_tensors call pattern.

  Moved AWQ Triton helpers into quantization/awq/awq_triton.py and removed the old top-level quantization/awq_triton.py.

  Split backend-specific kernel logic into:
  - hardware_backend/gpu/quantization/awq_kernels.py
  - hardware_backend/npu/quantization/awq_kernels.py

  This keeps awq.py focused on config, method dispatch, and scheme selection, while concrete weight handling and execution live in schemes and backend kernels.

## Accuracy Tests

GPU tests:
image
NPU tests:
image

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/srt/hardware_backend/npu/quantization/awq_kernels.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+            w13_offset=layer.w13_qzeros,
+            w2=layer.w2_qweight,
+            w2_scale=layer.w2_scales,
+            w2_offset=layer.w2_qzeros,
+            topk_weights=topk_weights,
+            topk_ids=topk_ids,
+            top_k=topk_ids.shape[1],
+            use_wna16=True,
+        )
+        return StandardCombineInput(hidden_states=output)
```
Comment (TamirBaydasov) [inline python/sglang/srt/hardware_backend/npu/quantization/awq_kernels.py]: Since we are not gated behind specific kernel implementations,  could you look if it is possible to call `NPUW4A16Int4DynamicMoEMethod` as a kernel here?
Example from one of our MoE refactoring PRs: https://github.com/sgl-project/sglang/pull/17361/changes#diff-34cc9aacc2ffaa0ad8351300aad66099bcbc2451d9a0a2c089aab5926d4f5e01
It should work for both apply and process_weights.

File: python/sglang/srt/layers/quantization/awq/schemes/awq_w4a16_moe.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+            AWQAscendMoEKernel,
+        )
+
+        self.kernel = AWQAscendMoEKernel(quant_config)
+
+    def create_moe_runner(
+        self, layer: torch.nn.Module, moe_runner_config: MoeRunnerConfig
+    ):
+        self.moe_runner_config = moe_runner_config
+        self.kernel.moe_runner_config = moe_runner_config
```
Comment (TamirBaydasov) [inline python/sglang/srt/layers/quantization/awq/schemes/awq_w4a16_moe.py]: Can we merge `awq_moe.py` file and this file together like it's done for Linear schemes in `awq_w4a16.py`?

**Review by b8zhong — COMMENTED** [review]
Hi, can you leave the GPU code outside of hardware backend? Since I think this is designed for other hardware backends, if I'm not misunderstanding, right?

File: python/sglang/srt/hardware_backend/gpu/quantization/awq_kernels.py [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+if TYPE_CHECKING:
+    from sglang.srt.layers.moe.token_dispatcher import (
+        CombineInput,
+        StandardDispatchOutput,
+    )
+    from sglang.srt.layers.quantization.base_config import QuantizationConfig
+
+_is_cuda = is_cuda()
+_is_hip = is_hip()
+_is_xpu = is_xpu()
```
Comment (ping1jing2) [inline python/sglang/srt/hardware_backend/gpu/quantization/awq_kernels.py]: if you've a seperate gpu folder, i think we should reduce or delete all `is_xxx` code, right?

File: python/sglang/srt/hardware_backend/npu/quantization/awq_kernels.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+        qzeros_tmp = layer.qzeros.data
+        qzeros_list = []
+        shifts = [0, 4, 1, 5, 2, 6, 3, 7]
+
+        for i in range(0, self.quant_config.pack_factor):
+            shift_num = shifts[i] * 4
+            qzeros_list.append((qzeros_tmp.reshape(-1, 1) >> shift_num) & 0xF)
+            qweight_tmp.bitwise_or_(
+                ((layer.qweight.data >> shift_num) * (2 ** (4 * i))) & (0xF << (4 * i))
+            )
```
[the remark this answers is not in the document]
Comment (ping1jing2) [reply python/sglang/srt/hardware_backend/npu/quantization/awq_kernels.py:35]: make sense

File: python/sglang/srt/layers/quantization/awq/schemes/awq_moe.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+__all__ = ["AWQMoEScheme"]
+
+
+class AWQMoEScheme(AWQMoESchemeBase):
+    def __init__(self, quant_config: "AWQMarlinConfig"):
+        self.quant_config = quant_config
+        self.kernel = AWQMoEKernel(quant_config)
+        if self.quant_config.weight_bits != 4:
+            raise ValueError("AWQMoEMethod only supports 4bit now.")
```
[the remark this answers is not in the document]
Comment (ping1jing2) [reply python/sglang/srt/layers/quantization/awq/schemes/awq_moe.py]: make sense

**Comment by OrangeRedeng** [comment]
LGTM. Maybe we should to rename awq_w4a16.py -> awq_linear.py to make it clearer

**Comment by Alisehen** [comment]
> LGTM. Maybe we should to rename awq_w4a16.py -> awq_linear.py to make it clearer

 Thanks, renamed `awq_w4a16.py` to `awq_linear.py` and updated the corresponding import in `schemes/__init__.py`.

**Comment by TamirBaydasov** [comment]
LGTM

**Comment by ping1jing2** [comment]
https://github.com/sgl-project/sglang/actions/runs/24952160506/job/73145780098?pr=21126

File: python/sglang/srt/hardware_backend/gpu/quantization/awq_kernels.py [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+except ImportError:
+    try:
+        from sglang.srt.layers.quantization.awq.awq_triton import (
+            awq_dequantize_triton as awq_dequantize,
+        )
+    except ImportError:
+        try:
+            from sgl_kernel import awq_dequantize
+        except ImportError:
+            pass
```
Comment (alexnails) [inline python/sglang/srt/hardware_backend/gpu/quantization/awq_kernels.py]: I believe there is now a regression for XPU here? Since it will go directly to triton

File: python/sglang/srt/layers/quantization/awq/schemes/awq_moe.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        self,
+        layer: torch.nn.Module,
+        dispatch_output: "StandardDispatchOutput",
+    ):
+        return self.kernel.apply(layer, dispatch_output)
+
+
+class AWQAscendMoEScheme(AWQMoEScheme):
+    def __init__(self, quant_config: "AWQConfig"):
+        super().__init__(quant_config)
```
Comment (alexnails) [inline python/sglang/srt/layers/quantization/awq/schemes/awq_moe.py]: `super().__init__(quant_config)` has `AWQMoEKernel` for GPU, which transitively imports `MarlinMoeQuantInfo` and `marlin_utils`. Not needed for NPU.

Suggest skipping the parent init or refactoring to a `_init_kernel()` hook:

```python
class AWQAscendMoEScheme(AWQMoEScheme):
    def __init__(self, quant_config: "AWQConfig"):
        # skip AWQMoEScheme.__init__
        from sglang.srt.hardware_backend.npu.quantization.awq_kernels import (
            AWQAscendMoEKernel,
        )
        self.quant_config = quant_config
        if self.quant_config.weight_bits != 4:
            raise ValueError("AWQAscendMoEScheme only supports 4bit now.")
        self.kernel = AWQAscendMoEKernel(quant_config)
```

Let's talk about this a bit further as I think the cleaner long-term fix is making `self.kernel` come from a platform factory (see plugin integration note on `awq.py`), at which point this subclass disappears entirely.

File: python/sglang/srt/layers/quantization/awq/schemes/awq_cpu.py [hunk 8]
```diff
@@ -1,65 +1,31 @@
...
-            return AWQMoEIntelAMXMethod(self)
-        return None
 
+class AWQIntelAMXLinearScheme(AWQLinearScheme):
+    """Linear scheme for AWQ on Intel CPU with AMX."""
 
-class AWQLinearIntelAMXMethod(AWQLinearMethod):
-    """Linear method for AWQ on Intel CPU with AMX."""
+    def __init__(self, quant_config: "AWQConfig"):
+        self.quant_config = quant_config
```
Comment (alexnails) [inline python/sglang/srt/layers/quantization/awq/schemes/awq_cpu.py:26]: `AWQIntelAMXLinearScheme` overrides `__init__` but doesn't call `super().__init__()` and doesn't set `self.kernel`.  Follows other comment but lets clean this up or make it less brittle?

File: python/sglang/srt/layers/linear.py [hunk 9]
```diff
@@ -54,10 +54,8 @@
 
 WEIGHT_LOADER_V2_SUPPORTED = [
     "CompressedTensorsLinearMethod",
-    "AWQMarlinLinearMethod",
     "AWQLinearMethod",
     "AWQLinearAscendMethod",
```
Comment (alexnails) [inline python/sglang/srt/layers/linear.py]: `AWQLinearAscendMethod` was deleted in this PR --> Ascend now goes through the unified `AWQLinearMethod` + `AWQAscendLinearScheme`. Since `WEIGHT_LOADER_V2_SUPPORTED` is matched by class name string, we should rm?

File: python/sglang/srt/layers/quantization/awq/awq.py [hunk 10]
```diff
@@ -0,0 +1,481 @@
...
+        return AWQLinearScheme(self)
+
+    def get_moe_scheme(self, layer: torch.nn.Module):
+        from sglang.srt.layers.moe.fused_moe_triton import FusedMoE
+
+        assert isinstance(layer, FusedMoE)
+        if _is_npu:
+            return AWQAscendMoEScheme(self)
+        raise NotImplementedError("AWQConfig only supports MoE scheme on NPU.")
+
```
Comment (alexnails) [inline python/sglang/srt/layers/quantization/awq/awq.py:175]: **Plugin-integration note (#21388 follow-up).**

TODO needed here for Multiplatform plugin. (even if you just mark with need to integrate `current_platform.is_out_of_tree`

something like: 

```python
def get_linear_scheme(self, layer: torch.nn.Module):
    assert isinstance(layer, LinearBase)
    from sglang.srt.platforms import current_platform
    cls = current_platform.get_awq_linear_scheme_cls()
    if cls is not None:
        return cls(self)
    return AWQLinearScheme(self)  # in-tree CUDA default

def get_moe_scheme(self, layer: torch.nn.Module):
    from sglang.srt.platforms import current_platform
    cls = current_platform.get_awq_moe_scheme_cls()
    if cls is None:
        raise NotImplementedError(
            f"AWQ MoE not provided by platform {current_platform.get_dispatch_key_name()!r}."
        )
    return cls(self)
```

With `SRTPlatform` extended to expose `get_awq_linear_scheme_cls()` / `get_awq_moe_scheme_cls()` / `get_awq_marlin_linear_scheme_cls()` (returning `None` by default; concrete platforms override). This matches how PR #21388 already exposes `get_mha_kv_pool_cls()`, `get_graph_runner_cls()`, etc.

Another option is to push the platform factory down to the `.kernel` layer (`AWQLinearScheme.__init__` calls `current_platform.get_awq_linear_kernel_cls()`). This would eliminate `AWQAscendLinearScheme` / `AWQIntelAMXLinearScheme` as subclasses entirely since they become kernel registrations on the OOT platform plugin. Also fixes the `super().__init__` side-effect issues I mentioned earlier

File: python/sglang/srt/layers/quantization/awq/awq.py [hunk 11]
```diff
@@ -0,0 +1,481 @@
...
+            return AWQLinearMethod(self)
+        return None
+
+    def get_linear_scheme(self, layer: torch.nn.Module):
+        assert isinstance(layer, LinearBase)
+        if _is_npu:
+            return AWQAscendLinearScheme(self)
+        return AWQLinearScheme(self)
+
+    def get_moe_scheme(self, layer: torch.nn.Module):
```
Comment (alexnails) [inline python/sglang/srt/layers/quantization/awq/awq.py:167]: nit: `get_moe_scheme` raising `NotImplementedError` for non-NPU is unreachable today (caller returns `None` for FusedMoE on non-NPU before consulting it). Either remove the raise or document the intent.

File: python/sglang/srt/layers/quantization/awq/awq.py (hunk 10 above)
Comment (Alisehen) [reply python/sglang/srt/layers/quantization/awq/awq.py:175]: Thanks for the review. I addressed the AWQ scheme issues:

  - Restored the XPU AWQ dequant path to use `sgl_kernel.awq_dequantize` instead of falling through to Triton.
  - Refactored AWQ Linear/MoE schemes to use `_init_kernel()` hooks, so Ascend no longer initializes the default GPU/Marlin kernel before replacing it.
  - Updated the CPU AMX AWQ path to use CPU-specific kernel objects behind the scheme, avoiding the brittle subclass-without-`super().__init__()` pattern.
  - Removed the stale `AWQLinearAscendMethod` entry from `WEIGHT_LOADER_V2_SUPPORTED`.
  - Added a TODO for moving AWQ scheme/kernel selection into the multiplatform plugin factory once quantization hooks are available.

**Comment by ping1jing2** [comment]
please confirm this is flasky ut or code accuracy issue
https://github.com/sgl-project/sglang/actions/runs/25057955960/job/73404880451?pr=21126#step:7:1108

**Review by ping1jing2 — APPROVED** [review]

**Comment by ping1jing2** [comment]
I merged it since several committer already reviewed and we confirmed that only one GPU failed CI is unrelated to our change
